File: Backtracking/WordBreakII.py
from typing import List
from collections import defaultdict
# from functools import lru_cache

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> List[str]:
        mem = defaultdict(list)
        # for w in wordDict:
        #     mem[w].append(w)
        return self.backtrack(s, set(wordDict), mem)

    def backtrack(self, s, wordset, mem):
        if s in mem:
            return mem[s]

        res = []
        if s in wordset:
            res.append(s)
        for w in wordset:
            if s.startswith(w):
                res += [w + ' ' + post for post in self.backtrack(s[len(w):], wordset, mem)]
        mem[s] = res
        return res


if __name__ == '__main__':
    solu = Solution()
    # s = "catsanddogsand"
    # wordDict = ["cat", "cats", "and", "sand", "dog", "dogs"]
    # s = "pineapplepenapple"
    # wordDict = ["apple", "pen", "applepen", "pine", "pineapple"]
    s = "aaaaaaa"
    wordDict = ["aaaa","aa","a"]
    res = solu.wordBreak(s, wordDict)
    print(res)

File: Backtracking/WordBreakII.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <unordered_set>

using std::string;
using std::vector;
using std::unordered_map;
using std::unordered_set;

class Solution {
    unordered_map<int, vector<string>> mem;
    unordered_set<string> wordSet;
public:
    vector<string> wordBreak(string s, vector<string>& wordDict) {
        wordSet = unordered_set(wordDict.begin(), wordDict.end());
        int n = s.size();
        mem[n] = {""};
        vector<string> res = backtrack(s, 0, n);
        return res;
    }

    vector<string> backtrack(const string& s, int idx, int n) {
        if (mem.count(idx) > 0) {
            return mem[idx];
        }
        vector<string> res;
        for (int i = idx+1; i <= n; i++) {
            string w = s.substr(idx, i - idx);
            if (wordSet.count(w) > 0) {
                vector<string> nxt = backtrack(s, i, n);
                for (string& sub_s : nxt) 
                    res.emplace_back(sub_s.empty() ? w : w + " " + sub_s);
            }
        }
        mem[idx] = res;
        return res;
    }
};

int main()
{
    Solution solu;
    string s("catsanddog");
    vector<string> wordDict {"cats", "cat", "and", "sand", "dog"};
    vector<string> ans = solu.wordBreak(s, wordDict);
    for (string& sub_s : ans)
        std::cout << sub_s << '\n';
    return 0;
}